### ISSUE TITLE
Using AP4_DefaultAtomFactory::Instance is not thread safe

### ISSUE THREAD

**Issue by boringuy** [axiomatic-systems/Bento4#64]
There are several places in the lib with hardcoded AP4_DefaultAtomFactory::Instance without a way to change the AtomFactory used to parse the atoms. For example:

In AP4_Atom::Clone(),
`AP4_DefaultAtomFactory::Instance.CreateAtomFromStream(*mbs, clone);`

Specifically, if the CreateAtomFromStream() happens to involve PushContext() in this static instance of the AtomFactory object, you will get random UnknownSampleEntry or other Unknown atom because of the change in the m_ContextStack. This will cause race condition when using the library in a multi-threaded environment. One way to let the caller pass in a reference to a AtomFactory like the AP4_File is better. 

There are other places with similar issue like the AP4_LinearReader, AP4_MarlinIpmpEncryptingProcessor ...etc.

Andy

**Comment by barbibulle** [comment]
You're right, this is something that should be fixed. I'll try to include a fix for it in the next release.

**Comment by baconz** [comment]
It looks like you fixed this in 2e3fb80a, but you left the `AP4_DefaultAtomFactory::Instance_` as the default atom factory for `Processor->Process()`. This tripped me up pretty hard.

Would you be OK with breaking backwards compatibility and instantiating a new atom factory for each call to `Process`? It doesn't look like this would break any of the Apps.

**Comment by barbibulle** [comment]
I think that would be a good thing to do.
